src/Type/TypeCombinator.php

```diff
@@ -833,16 +842,21 @@ private static function countConstantArrayValueTypes(array $types): int
 	}
 
 	/**
-	 * @param Type[] $constantArrays
-	 * @return Type[]
+	 * @param list<Type> $constantArrays
+	 * @return list<Type>
 	 */
-	private static function reduceArrays(array $constantArrays): array
+	private static function reduceArrays(array $constantArrays, bool $preserveTaggedUnions): array
 	{
 		$newArrays = [];
 		$arraysToProcess = [];
 		$emptyArray = null;
 		foreach ($constantArrays as $constantArray) {
 			if (!$constantArray->isConstantArray()->yes()) {
+				// This is an optimization for current use-case of $preserveTaggedUnions=false, where we need
+				// one constant array as a result, or we generalize the $constantArrays.
+				if (!$preserveTaggedUnions) {
+					return $constantArrays;
+				}
```

Comment:
I'm conflicted about this. On one hand, it would feel bad not to have this condition and run a lot of the method uselessly. But on the other hand, the condition depends on the particular usage of `$preserveTaggedUnions`. Maybe a better name for `$preserveTaggedUnions` would help, but I didn't come up with anything. So I at least added a comment.

Reply:
This is fine 👍